syncapi/storage/shared/syncserver.go

```diff
@@ -563,21 +567,69 @@ func (d *Database) addTypingDeltaToResponse(
 	return nil
 }
 
+// addReceiptDeltaToResponse adds all receipt information to a sync response
+// since the specified position
+func (d *Database) addReceiptDeltaToResponse(
+	since types.StreamingToken,
+	joinedRoomIDs []string,
+	res *types.Response,
+) error {
+	var jr types.JoinResponse
+	// check all joinedRooms for receipts
+	for _, roomID := range joinedRoomIDs {
+		receipts, err := d.Receipts.SelectRoomReceiptsAfter(context.TODO(), roomID, since.EDUPosition())
+		if err != nil {
+			return err
+		}
+
+		ev := gomatrixserverlib.ClientEvent{
+			Type:   "m.receipt",
+			RoomID: roomID,
+		}
+		content := map[string]interface{}{}
+		for _, receipt := range receipts {
+			content[receipt.EventID] = map[string]interface{}{
+				"m.read": map[string]interface{}{
+					receipt.UserID: struct {
+						gomatrixserverlib.Timestamp `json:"ts"`
+					}{
+						receipt.Timestamp,
+					},
+				},
+			}
+
+			ev.Content, err = json.Marshal(content)
+			if err != nil {
+				return err
+			}
+			jr.Ephemeral.Events = append(jr.Ephemeral.Events, ev)
+			res.Rooms.Join[roomID] = jr
```

Comment:
What happens if there is already a `JoinResponse` in `res`? This will replace it which doesn't sound right?

Reply:
Not yet sure if the solution to this will work properly.